### PR TITLE
Fixes #29288 - reorder Config management settings

### PR DESCRIPTION
**Pull request by ezr-ondrej**
Settings that were previously related only to puppet, but are now used
in all CfgMgmt tools so this creates new setting category.

Settings moved from Puppet to Cfgmgmt:
- `matchers_inheritance`
- `Default_parameters_Lookup_Path`
- `interpolate_erb_in_parameters`
- `update_subnets_from_facts`
- `update_hostgroup_from_facts`
- `create_new_host_when_facts_are_uploaded`
- `create_new_host_when_report_is_uploaded`
- `location_fact`
- `organization_fact`
- `default_location`
- `default_organization`
- `always_show_configuration_status`

from Provisioning to Cfgmgmt:
- `ignore_facts_for_operatingsystem`
- `ignore_facts_for_domain`
- `ignored_interface_identifiers`
- `excluded_facts`

from Puppet to Provisioning:
- `use_uuid_for_certificates`

from Provisioning to Puppet:
- `ignore_puppet_facts_for_provisioning`

### CODE REVIEW

**Comment by lzap** [comment]
Shall we create a new common "Configuration" group for these? @ares @ekohl

**Comment by ares** [comment]
I like that, just the wording may be a bit misleading. Having a category Configuration under Setting sounds weird to me. Perhaps Configuration Management? @ezr-ondrej is that feasible or does it make the PR much more complicated?

**Comment by ezr-ondrej** [comment]
> I like that, just the wording may be a bit misleading. Having a category Configuration under Setting sounds weird to me. Perhaps Configuration Management? @ezr-ondrej is that feasible or does it make the PR much more complicated?

Nope, that definitely makes sense to me :+1:
I agree that Configuration might be missleading, so I'll start with `Configuration Management`, but I feel it is quite long, so ideas are welcome ;)

File: app/models/setting/cfgmgmt.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      set('organization_fact', N_("Hosts created from uploaded facts will be placed in the organization this fact dictates. The content of this fact should be the full label of the organization."), 'foreman_organization', N_('Organization fact')),
+      set('update_subnets_from_facts', N_("Foreman will update a Host's subnet from its facts"), 'none', N_('Update subnets from facts'), nil, { :collection => proc { update_subnets_from_facts_types } }),
+      set('update_hostgroup_from_facts', N_("Foreman will update a Host's Hostgroup from its facts"), true, N_('Update hostgroup from facts')),
+      set('ignored_interface_identifiers', N_("Ignore interfaces that match these values during facts importing, you can use * wildcard to match names with indexes e.g. macvtap*"), IGNORED_INTERFACES, N_('Ignore interfaces with matching identifier')),
+      set('always_show_configuration_status', N_("All hosts will show a configuration status even when there is no configuration report for the Host"), false, N_('Always show configuration status')),
+    ]
+  end
+
+  def self.humanized_category
+    N_('Configuration management')
```
Comment (ezr-ondrej) [inline app/models/setting/cfgmgmt.rb:27]: Ideas for a better name? :thinking:

Comment (ares) [reply app/models/setting/cfgmgmt.rb:27]: Config management - I prefer this, tried locally and seemed OK
Config mgmt - not great not terrible
Cfgmgmt - hopefully not

Comment (ekohl) [reply app/models/setting/cfgmgmt.rb:27]: Let's start with a real human readable version (as you have it now). If we run into troubles, we can always revisit it.

File: app/models/setting/cfgmgmt.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      set('default_organization', N_("Hosts created from uploaded facts that did not send a organization fact will be placed in this organization"), '', N_('Default organization'), nil, {:collection => proc { Hash[Organization.all.map { |org| [org[:title], org[:title]] }] } }),
+      set('location_fact', N_("Hosts created from uploaded facts will be placed in the location this fact dictates. The content of this fact should be the full label of the location."), 'foreman_location', N_('Location fact')),
+      set('organization_fact', N_("Hosts created from uploaded facts will be placed in the organization this fact dictates. The content of this fact should be the full label of the organization."), 'foreman_organization', N_('Organization fact')),
+      set('update_subnets_from_facts', N_("Foreman will update a Host's subnet from its facts"), 'none', N_('Update subnets from facts'), nil, { :collection => proc { update_subnets_from_facts_types } }),
```
Comment (ekohl) [inline app/models/setting/cfgmgmt.rb]: Why is this a proc if the values are static? Can't this be a constant? Or does that break translations?

File: app/models/setting/provisioning.rb [hunk 3]
```diff
@@ -58,8 +58,6 @@ def self.default_settings
       set('safemode_render', N_("Enable safe mode config templates rendering (recommended)"), true, N_('Safemode rendering')),
       set('access_unattended_without_build', N_("Allow access to unattended URLs without build mode being used"), false, N_('Access unattended without build')),
       set('manage_puppetca', N_("Foreman will automate certificate signing upon provision of new host"), true, N_('Manage PuppetCA')),
-      set('ignore_puppet_facts_for_provisioning', N_("Stop updating IP address and MAC values from Puppet facts (affects all interfaces)"), false, N_('Ignore Puppet facts for provisioning')),
-      set('ignored_interface_identifiers', N_("Ignore interfaces that match these values during facts importing, you can use * wildcard to match names with indexes e.g. macvtap*"), IGNORED_INTERFACES, N_('Ignore interfaces with matching identifier')),
       set('ignore_facts_for_operatingsystem', N_("Stop updating Operating System from facts"), false, N_('Ignore facts for operating system')),
       set('ignore_facts_for_domain', N_("Stop updating domain values from facts"), false, N_('Ignore facts for domain')),
```
Comment (ekohl) [inline app/models/setting/provisioning.rb]: Shouldn't these also be moved?

File: app/models/setting/cfgmgmt.rb (hunk 2 above)
Comment (ezr-ondrej) [reply app/models/setting/cfgmgmt.rb]: Exactly it needs to be evaluated on request time as the translation needs to know what language to use.

**Review by tbrisker — COMMENTED** [review]
should we split between facts and config management category? since facts can also come for example from discovery or sub-man.

**Comment by ekohl** [comment]
I think that makes sense

**Comment by ezr-ondrej** [comment]
Hmm it make sense, but I'd rather find a unifying name then creating two categories if possible, as we already have quite a lot with just few settings each and Users who don't visit often might get quite overwhelmed especially with common (REX, Ansible/Puppet, discovery) plugins installed.

What about `Facts and Parameters`?

File: app/models/setting/cfgmgmt.rb (hunk 1 above)
Comment (ezr-ondrej) [reply app/models/setting/cfgmgmt.rb:27]: As per disscussion below what about What about `Facts and Parameters`?

**Comment by ekohl** [comment]
I don't think too many categories is a huge concern now. I wouldn't mind it if we had subcategories, so you could end up with
* Configuration Management
  * Common
  * Ansible (provided by foreman_ansible)
  * Puppet (provided by foreman_puppet)

However, that's probably too much to ask for this PR.

File: app/models/setting/cfgmgmt.rb (hunk 1 above)
Comment (ekohl) [reply app/models/setting/cfgmgmt.rb:27]: No, I think that's bad. If I wanted to look for Puppet options, I wouldn't look at `Facts and Parameters`.

**Review by tbrisker — COMMENTED** [review]
Would it perhaps make sense to have `Facts` and `Config management` categories, and completely remove the puppet category? Then in the puppet plugin, add the setting currently in Puppet that only make sense with the plugin to one of the two?

**Comment by ekohl** [comment]
> Would it perhaps make sense to have `Facts` and `Config management` categories, and completely remove the puppet category? Then in the puppet plugin, add the setting currently in Puppet that only make sense with the plugin to one of the two?

I wasn't explicit in that, but I was indeed under the impression that that was the proposal. I'd even consider the Puppet plugin would add its settings to the Configuration Management category.

**Comment by ezr-ondrej** [comment]
> Would it perhaps make sense to have `Facts` and `Config management` categories, and completely remove the puppet category? Then in the puppet plugin, add the setting currently in Puppet that only make sense with the plugin to one of the two?

Sounds good :+1: I'll do so

File: app/models/setting/puppet.rb [hunk 4]
```diff
@@ -1,32 +1,12 @@
 class Setting::Puppet < Setting
-  def self.update_subnets_from_facts_types
-    {
-      'all' => _('All'),
-      'provisioning' => _('Provisioning only'),
-      'none' => _('None'),
-    }
-  end
-
   def self.default_settings
```
Comment (tbrisker) [inline app/models/setting/puppet.rb]: Shouldn't the remaining settings also move to `config management` category, perhaps as a separate method that will be extracted to the plugin? (maybe with the exception of `update_environment_from_facts` which should go to facts)

File: app/models/setting/cfgmgmt.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+class Setting::Cfgmgmt < Setting
+  def self.default_settings
+    [
+      set('matchers_inheritance', N_("Foreman matchers will be inherited by children when evaluating Lookups for Hostgroups, Organizations and Locations"), true, N_('Matchers inheritance')),
+      set('Default_parameters_Lookup_Path', N_("Foreman will evaluate host lookups in this order by default"), ["fqdn", "hostgroup", "os", "domain"], N_('Default parameters lookup path')),
+      set('interpolate_erb_in_parameters', N_("Foreman will parse ERB in parameters and lookup values"), true, N_('Interpolate ERB in parameters')),
+      set('create_new_host_when_report_is_uploaded', N_("Foreman will create the host when a report is received"), true, N_('Create new host when report is uploaded')),
```
Comment (tbrisker) [inline app/models/setting/cfgmgmt.rb]:
```suggestion
      set('create_new_host_when_report_is_uploaded', N_("Foreman will create the host when a configuration report is received"), true, N_('Create new host when report is uploaded')),
```

**Review by tbrisker — CHANGES_REQUESTED** [review]
I think you forgot to commit `facts.rb`

File: app/models/setting/puppet.rb (hunk 4 above)
Comment (ezr-ondrej) [reply app/models/setting/puppet.rb]: eeeeeeh, well probably, is it a proposal for moving Ansible settings to config management as well?

Comment (tbrisker) [reply app/models/setting/puppet.rb]: probably at least those that relate to using ansible for config management (e.g. out of sync time etc), perhaps those that relate to ansible as a rex provider should be in the rex settings

File: app/models/setting/cfgmgmt.rb (hunk 1 above)
Comment (ezr-ondrej) [reply app/models/setting/cfgmgmt.rb:27]: What about now? We have two categories `Facts` and `Configuration management` and both support "subcategories" (I'd extract that to a module, but just if I've understood correctly what we want)

File: app/models/setting/cfgmgmt.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
+class Setting::Cfgmgmt < Setting
+  def self.tool_specific
+    @tool_specific ||= {}
+  end
+
+  def self.add_tool_specific(tool_name, settings)
+    tool_specific[tool_name.to_sym] ||= []
+    tool_specific[tool_name.to_sym].concat(settings)
+  end
```
Comment (tbrisker) [inline app/models/setting/cfgmgmt.rb:9]: Would it make sense to have this on the base setting (with another option for category) to enable easy extension of existing categories from plugins?

Comment (ezr-ondrej) [reply app/models/setting/cfgmgmt.rb:9]: I guess... But in that case, I'd prefer to finish the setting register and do it there.
As it is getting out of hand for this PR IMHO

**Comment by ezr-ondrej** [comment]
Closing in favor of #8874

**Pull request closed without merging**